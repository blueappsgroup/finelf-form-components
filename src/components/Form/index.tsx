import React, { FC, ReactElement, useMemo, useState } from 'react'
import { Formik, Form } from 'formik'
import styled from 'styled-components'

import { ThemeProvider } from '../../consts/theme'
import { device } from '../../consts/sizes'
import { FormProps, FormValuesType } from '../../types'
import RedirectPage from '../RedirectPage'
import {
  FormContext,
  getFormValuesFromCache,
  resetFormValueCache,
  handleSendDataToApi,
  getFieldsValuesFromUrl,
} from '../../utils'
import { formStatuses } from '../../consts/form'

const StyledForm = styled(Form)`
  display: flex;
  flex-direction: column;
  max-width: ${(props): string => props.theme.formMaxWidth};
  justify-self: center;
  background: ${(props): string => props.theme.formBgColor};
  font-family: ${(props): string => props.theme.fontFamily};
  border-radius: 6px;
  box-shadow: ${(props): string => props.theme.formBoxShadow};
`

const FormWrapper: FC<FormProps> = ({
  children,
  onSubmit,
  customTheme,
  id,
  stepsLength,
  stepsTitles,
  hasRedirect,
  redirectUrl,
  timeToRedirect,
  redirectHeaderText,
  logoImg,
  redirectMainImg,
  redirectBgImg,
  sendDataToApi,
  apiUrl,
  transactionName,
  propertyNamesFromUrl,
  dataWithUserAgent,
}) => {
  const trasationIdValue =
    transactionName &&
    new URLSearchParams(window.location.search).get(transactionName)
  const intialValuesFromUrl = useMemo(
    () =>
      (propertyNamesFromUrl &&
        propertyNamesFromUrl.length > 0 &&
        getFieldsValuesFromUrl(propertyNamesFromUrl)) ||
      {},
    [propertyNamesFromUrl]
  )
  const [redirectUrlPath, setRedirectUrlPath] = useState(redirectUrl)
  const [initialValues, setInitialValues] = useState({
    ...intialValuesFromUrl,
    ...getFormValuesFromCache(id),
    // eslint-disable-next-line @typescript-eslint/camelcase
    trasaction_id: trasationIdValue,
  })
  const [currentStep, setCurrentStep] = useState(0)
  const [fieldsForSkip, setFieldsForSkip] = useState<string[]>([])
  const [errorFromApi, setErrorFromApi] = useState<boolean>(false)
  const shouldRedirect = !errorFromApi && hasRedirect
  const addFieldForSkip = (key: string): void =>
    setFieldsForSkip([...fieldsForSkip, key])

  const handleSubmit = async (
    values: FormValuesType,
    props: {
      resetForm: () => void
      setStatus: (status: string) => void
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ): Promise<any> => {
    if (onSubmit) {
      onSubmit(values, props)
    }

    try {
      if (sendDataToApi && apiUrl) {
        const response = await handleSendDataToApi(
          values,
          apiUrl,
          id,
          fieldsForSkip,
          dataWithUserAgent
        )
        const {
          redirectUrl: urlFromApi,
          status: statusFromApi,
        } = await response.json()

        if (statusFromApi === false) {
          setErrorFromApi(true)
          return
        }

        urlFromApi && setRedirectUrlPath(urlFromApi)
      }

      props.resetForm()
      props.setStatus(formStatuses.submited)
    } catch (e) {
      props.setStatus(formStatuses.error)
    }
  }
  const handleReset = (): void => {
    resetFormValueCache(id)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    setInitialValues({} as any)
  }

  const prevStep: Function = () => setCurrentStep(currentStep - 1)

  const nextStep: Function = () => setCurrentStep(currentStep + 1)

  return (
    <FormContext.Provider
      value={{
        id,
        apiUrl,
        stepsLength,
        currentStep,
        stepsTitleList: stepsTitles,
        nextStep,
        prevStep,
        fieldsForSkip,
        addFieldForSkip,
        errorFromApi,
      }}
    >
      <ThemeProvider customTheme={{ ...customTheme }}>
        <Formik
          enableReinitialize
          initialValues={initialValues}
          onSubmit={handleSubmit}
          onReset={handleReset}
        >
          {(props): ReactElement =>
            (shouldRedirect && props.status === formStatuses.submited && (
              <RedirectPage
                redirectUrl={redirectUrlPath}
                backgroundImage={redirectBgImg}
                logoImg={logoImg}
                headerText={redirectHeaderText}
                timeToRedirect={timeToRedirect}
                mainImg={redirectMainImg}
              />
            )) || (
              <StyledForm id={id}>
                {children}
              </StyledForm>
            )
          }
        </Formik>
      </ThemeProvider>
    </FormContext.Provider>
  )
}

export default FormWrapper
