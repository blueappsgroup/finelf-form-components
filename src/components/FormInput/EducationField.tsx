import { Field } from 'formik'
import React, { ReactElement } from 'react'
import { SelectFieldProps } from '../../types'

import { BaseSelectField } from './base'
import { validateSelect } from './validate'

const options = {
  none: 'brak',
  primary: 'podstawowe',
  vocational: 'zawodowe',
  secondary: 'średnie',
  higher: 'wyższe licencjackie',
  masters: 'wyższe magisterskie',
  phd: 'wyższe doktorskie',
}

const EducationField: (props: SelectFieldProps) => ReactElement = ({
  id,
  label,
  name = 'education',
  required = false,
  validate,
  showError,
}) => (
  <Field
    required={required}
    showError={showError}
    id={id}
    name={name}
    component={BaseSelectField}
    label={label}
    type="select"
    options={options}
    validate={
      (validate && validate(required)) || validateSelect(required, options)
    }
  />
)

export default EducationField
