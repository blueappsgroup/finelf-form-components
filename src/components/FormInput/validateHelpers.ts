export const isEmpty = (value: string): boolean => !value

export const isNotValidLength = (value: string, from = 2, to = 100): boolean =>
  value.length < from || value.length > to

export const isNotValidPhoneNumber = (value: string): boolean =>
  !/^(45|50|51|53|57|60|66|69|72|73|78|79|88)[0-9]{7}$/.test(value)

export const isNotValidEmail = (value: string): boolean =>
  !/[a-zA-Z0-9_.+-]+@[a-zAZ0-9-]+\.[a-zA-Z0-9-.]+$/.test(value)

export const isNotValidOtherLoanAmount = (value: string): boolean =>
  !/^[0-9]{1,10}$/.test(value)
