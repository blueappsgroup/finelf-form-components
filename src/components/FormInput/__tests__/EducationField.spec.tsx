import * as React from 'react'
import { render, fireEvent, act } from '@testing-library/react'

import Form from '../../Form'
import EducationField from '../EducationField'
import { isNotValidOption } from '../validateHelpers'

describe('<EducationField />', () => {
  const onSubmit = jest.fn()
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const setupWrapper: (config: any) => any = ({
    formId = 'testForm',
    selectId = 'educationField',
    selectName = 'name',
    required = false,
    ...rest
  }) => {
    const wrapper = render(
      <Form id={formId} onSubmit={onSubmit}>
        <EducationField
          id={selectId}
          name={selectName}
          required={required}
          {...rest}
        />
      </Form>
    )
    const { container } = wrapper

    const select = container.querySelector(`[name="${selectName}"]`)

    return {
      select,
      ...wrapper,
    }
  }

  it('matches snapshot', () => {
    const { select } = setupWrapper({})

    expect(select).toBeTruthy()
    expect(select).toMatchSnapshot()
  })

  it('matches snapshot with label', () => {
    const wrapper = setupWrapper({ label: 'Test' })
    const { container } = wrapper

    expect(wrapper.getByText('Test')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it('matches snapshot with error', async () => {
    const wrapper = setupWrapper({
      label: 'Test',
      showError: true,
      required: true,
      error: 'true',
    })
    const { container, select } = wrapper

    await act(async () => {
      fireEvent.focus(select, { target: { value: 'primary' } })
    })

    await act(async () => {
      fireEvent.blur(select, { target: { value: 'primary' } })
    })

    expect(wrapper.getByText('To pole jest wymagane')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it('changes select field', async () => {
    const { select } = setupWrapper({})

    await act(async () => {
      fireEvent.change(select, { target: { value: 'masters' } })
    })

    expect(select.value).toBe('masters')
  })

  it('select field onBlur save value to sessionStorage', async () => {
    const { select } = setupWrapper({})

    await act(async () => {
      fireEvent.change(select, { target: { value: 'masters' } })
    })

    await act(async () => {
      fireEvent.blur(select, { target: { value: 'masters' } })
    })

    expect(
      JSON.parse(global.window.sessionStorage.getItem('form-testForm')).name
    ).toBe('masters')
    expect(select.value).toBe('masters')
  })

  it('select field with custom validation', async () => {
    const customValidate = () => (value: string): string => {
      if (
        isNotValidOption(value, {
          none: 'brak',
          primary: 'podstawowe',
          vocational: 'zawodowe',
          secondary: 'średnie',
          higher: 'wyższe licencjackie',
          masters: 'wyższe magisterskie',
          phd: 'wyższe doktorskie',
        })
      ) {
        return 'Podane dane są nieprawidłowe'
      }
    }
    const wrapper = setupWrapper({
      validate: customValidate,
      showError: true,
    })
    const { select, container } = wrapper

    await act(async () => {
      fireEvent.change(select, { target: { value: 'test' } })
    })

    await act(async () => {
      fireEvent.blur(select, { target: { value: 'test' } })
    })

    expect(container).toMatchSnapshot()
    expect(wrapper.getByText('Podane dane są nieprawidłowe')).toBeTruthy()
  })
})
