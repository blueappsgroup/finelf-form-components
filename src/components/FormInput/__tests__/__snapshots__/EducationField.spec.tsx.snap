// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EducationField /> matches snapshot 1`] = `
<select
  class="sc-iqAbSa hSYmLs"
  id="educationField"
  name="name"
  type="text"
>
  <option
    disabled=""
    value="select"
  >
    wybierz
  </option>
  <option
    value="none"
  >
    brak
  </option>
  <option
    value="primary"
  >
    podstawowe
  </option>
  <option
    value="vocational"
  >
    zawodowe
  </option>
  <option
    value="secondary"
  >
    średnie
  </option>
  <option
    value="higher"
  >
    wyższe licencjackie
  </option>
  <option
    value="masters"
  >
    wyższe magisterskie
  </option>
  <option
    value="phd"
  >
    wyższe doktorskie
  </option>
</select>
`;

exports[`<EducationField /> matches snapshot with error 1`] = `
<div>
  <form
    action="#"
    class="sc-iCoHVE gQzhKx"
    id="testForm"
  >
    <div
      class="sc-fujyUd curuhY"
    >
      <label
        for="name"
      >
        Test
      </label>
      <select
        class="sc-iqAbSa cTqqfZ"
        id="educationField"
        label="Test"
        name="name"
        required=""
        type="text"
      >
        <option
          disabled=""
          value="select"
        >
          wybierz
        </option>
        <option
          value="none"
        >
          brak
        </option>
        <option
          value="primary"
        >
          podstawowe
        </option>
        <option
          value="vocational"
        >
          zawodowe
        </option>
        <option
          value="secondary"
        >
          średnie
        </option>
        <option
          value="higher"
        >
          wyższe licencjackie
        </option>
        <option
          value="masters"
        >
          wyższe magisterskie
        </option>
        <option
          value="phd"
        >
          wyższe doktorskie
        </option>
      </select>
      <span
        class="sc-jrsJCI eQDYwE"
      >
        To pole jest wymagane
      </span>
    </div>
  </form>
</div>
`;

exports[`<EducationField /> matches snapshot with label 1`] = `
<div>
  <form
    action="#"
    class="sc-iCoHVE gQzhKx"
    id="testForm"
  >
    <div
      class="sc-fujyUd curuhY"
    >
      <label
        for="name"
      >
        Test
      </label>
      <select
        class="sc-iqAbSa hSYmLs"
        id="educationField"
        label="Test"
        name="name"
        type="text"
      >
        <option
          disabled=""
          value="select"
        >
          wybierz
        </option>
        <option
          value="none"
        >
          brak
        </option>
        <option
          value="primary"
        >
          podstawowe
        </option>
        <option
          value="vocational"
        >
          zawodowe
        </option>
        <option
          value="secondary"
        >
          średnie
        </option>
        <option
          value="higher"
        >
          wyższe licencjackie
        </option>
        <option
          value="masters"
        >
          wyższe magisterskie
        </option>
        <option
          value="phd"
        >
          wyższe doktorskie
        </option>
      </select>
    </div>
  </form>
</div>
`;

exports[`<EducationField /> select field with custom validation 1`] = `
<div>
  <form
    action="#"
    class="sc-iCoHVE gQzhKx"
    id="testForm"
  >
    <div
      class="sc-fujyUd curuhY"
    >
      <select
        class="sc-iqAbSa cTqqfZ"
        id="educationField"
        name="name"
        type="text"
      >
        <option
          disabled=""
          value="select"
        >
          wybierz
        </option>
        <option
          value="none"
        >
          brak
        </option>
        <option
          value="primary"
        >
          podstawowe
        </option>
        <option
          value="vocational"
        >
          zawodowe
        </option>
        <option
          value="secondary"
        >
          średnie
        </option>
        <option
          value="higher"
        >
          wyższe licencjackie
        </option>
        <option
          value="masters"
        >
          wyższe magisterskie
        </option>
        <option
          value="phd"
        >
          wyższe doktorskie
        </option>
      </select>
      <span
        class="sc-jrsJCI eQDYwE"
      >
        Podane dane są nieprawidłowe
      </span>
    </div>
  </form>
</div>
`;
